### PR TITLE
Improve deposit flow

### PR DESCRIPTION
**Pull request by michalsmiarowski**
Closes: #367 
Ref: #369 

We decided to change how the flow works:

**1.** User generates deposit address, download JSON, disconnects at step 2 and reconnects again (did not reveal the deposit yet)

**Solution:** Show him step 2

**2.** User generates deposit address, download JSON, sends the funds to the deposit address and then disconnects at step 3 and reconnects again (did not reveal the deposit yet)

**Solution:** Show him step 3

**3.** User generates deposit address, download JSON, sends the funds to the deposit address and then reveal the deposit and sees the success modal (step 4). He then disconnects at step 4 and reconnects again

**Solution:** Show him step 1 because he already revealed the deposit

**Additional notes:**

Do not allow the user to reveal multiple deposits that were sent to the same deposit address. The dApp should only reveal the oldest un-revealed deposit transaction (sent to this deposit address) after clicking the minting button at step 3.

**The back button:**

**1.** Clicking the back button on step 2 will display the New Generation address modal and take user to step 1

**2.** Clicking the back button on step 3 will display the New Generation address modal and the user to step 1 (he already send funds at this point so no need to navigate him back to step 2)

**3.** Clicking Dismiss or back button at step 4 will navigate the user to the step 1 without displaying the New Generation address modal

#### 1. Moving most of the logic in `MakeDeposit` component to `MintingFlowRouter`

We want to check if the funds are sent and if the utxo is revealed in the
component that is a parent to the minting steps component when the user
connects. This way we could display the proper step based on the actual data of
the user.

#### 2. Clearing the interval after the utxos are found

We clear the interval that search for the utxos after we found them in a given
deposit address. We won't need that interval anymore since we don't want to
promote sending multiple transactions to one deposit address. After we find one
we go with the flow and decide which step should be displayed to the user - if
it's not revealed then step 3. If it's revealed then it's step 4(TODO: change it
to step 1 + local storage data clean).

#### 3. Operate on only one UTXO instead of array of UTXOs

We will only operate on the oldest UTXO from the array returned from
`findAllUnspentTransactionOutputs`.

#### 5. Remove unnecessary updates of minting steps in the state

### CODE REVIEW

File: src/components/Modal/TbtcMintingConfirmationModal/index.tsx [hunk 1]
```diff
@@ -71,19 +68,12 @@ const TbtcMintingConfirmationModal: FC<TbtcMintingConfirmationModalProps> = ({
       refundLocktime,
     }
 
-    const successfulTransactions = await revealMultipleDeposits(
-      utxos,
-      depositScriptParameters
-    )
+    const transaction = await revealDeposit(utxo, depositScriptParameters)
```
Comment (r-czajkowski) [inline src/components/Modal/TbtcMintingConfirmationModal/index.tsx:71]: We can just `await revealDeposit(utxo, depositScriptParameters)`. We don't use `transaction` anywhere.

File: src/components/Modal/TbtcMintingConfirmationModal/index.tsx [hunk 2]
```diff
@@ -71,19 +68,12 @@ const TbtcMintingConfirmationModal: FC<TbtcMintingConfirmationModalProps> = ({
...
   }
 
   useEffect(() => {
     const getEstimatedFees = async () => {
-      const amount = utxos.reduce(
-        (accumulator, currentValue) =>
-          BigNumber.from(accumulator).add(currentValue.value),
-        BigNumber.from(0)
-      )
+      const amount = BigNumber.from(utxo.value)
```
Comment (r-czajkowski) [inline src/components/Modal/TbtcMintingConfirmationModal/index.tsx:76]: Let's extract `const amount = BigNumber.from(utxo.value)` outside the `useEffect` and add to dependency array(and other missing dependencies). See https://github.com/threshold-network/token-dashboard/pull/378/commits/4eb391a906b9a97bf5acd15e55a37273e55b219f

File: src/components/Modal/TbtcRecoveryFileModal/index.tsx [hunk 3]
```diff
@@ -45,7 +44,6 @@ const TbtcRecoveryFileModalModal: FC<
   const { updateState } = useTbtcState()
 
   const handleDoubleReject = () => {
-    updateState("mintingStep", MintingStep.Deposit)
```
Comment (r-czajkowski) [inline src/components/Modal/TbtcRecoveryFileModal/index.tsx:48]: Unused `const { updateState } = useTbtcState()` in line 44.

File: src/components/Modal/tBTC/GenerateNewDepositAddress.tsx [hunk 4]
```diff
@@ -28,7 +28,6 @@ const GenerateNewDepositAddressBase: FC<BaseModalProps> = ({ closeModal }) => {
 
   const onConfirmClick = () => {
     removeDepositDataFromLocalStorage()
-    resetDepositData()
```
Comment (r-czajkowski) [inline src/components/Modal/tBTC/GenerateNewDepositAddress.tsx:31]: Are we sure? Why don't we reset the deposit data once someone wants to generate a new deposit address?

Comment (r-czajkowski) [reply src/components/Modal/tBTC/GenerateNewDepositAddress.tsx:31]: If we are sure then let's remove ` const { resetDepositData } = useTbtcState()` and import `useTbtcState`.

File: src/components/Modal/TbtcRecoveryFileModal/index.tsx (hunk 3 above)
Comment (r-czajkowski) [reply src/components/Modal/TbtcRecoveryFileModal/index.tsx:48]: and let's remove import.

File: src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts [hunk 5]
```diff
@@ -36,6 +38,7 @@ export const useTBTCDepositDataFromLocalStorage = () => {
     }
     delete newLocalStorageData[`${account}`]
     setTBTCDepositData(newLocalStorageData)
+    resetDepositData() // additionaly remove the deposit data from the store
```
Comment (r-czajkowski) [inline src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts:41]: I would wrap those 2 hooks into a single one. IMO looks weird inside this `useTBTCDepositDataFromLocalStorage`. Eg
`useRemoveDepositData.`
```
const useRemoveDepositData() => {
  const { resetDepositData } = useTbtcState()
    const { removeDepositDataFromLocalStorage } =
    useTBTCDepositDataFromLocalStorage()
    
    return useCallback(() => {
      resetDepositData()
      removeDepositDataFromLocalStorage()
    })
}
```

Or remove deposit from local storage inside `useTbtcState` hook.

File: src/pages/tBTC/Bridge/MintingCard/InitiateMinting.tsx [hunk 6]
```diff
@@ -11,20 +11,19 @@ import withOnlyConnectedWallet from "../../../../components/withOnlyConnectedWal
 import { UnspentTransactionOutput } from "@keep-network/tbtc-v2.ts/dist/src/bitcoin"
 
 const InitiateMintingComponent: FC<{
-  utxos: UnspentTransactionOutput[] | undefined
+  utxo: UnspentTransactionOutput | undefined
```
Comment (r-czajkowski) [inline src/pages/tBTC/Bridge/MintingCard/InitiateMinting.tsx:14]: why `undefined` ? I think we should only redirect to step 3 if we find the utxo.

File: src/pages/tBTC/Bridge/MintingCard/MakeDeposit.tsx [hunk 7]
```diff
@@ -39,35 +39,10 @@ const AddressRow: FC<{ address: string; text: string }> = ({
 }
 
 const MakeDepositComponent: FC<{
-  utxos: UnspentTransactionOutput[] | undefined
```
Comment (r-czajkowski) [inline src/pages/tBTC/Bridge/MintingCard/MakeDeposit.tsx:42]: There are some unused imports in this file. Let's remove.

File: src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx [hunk 8]
```diff
@@ -41,7 +47,21 @@ const MintingFlowRouterBase = () => {
...
+          const utxo = utxos.pop()
+          setUtxo(utxo)
+          // If there is at least one utxo we remove the interval, because we
+          // don't want to encourage sending multiple transactions to the same
+          // deposit address.
+          clearInterval(interval)
+        } else {
+          // If ther is no utxo then we display the `MakeDeposit` step to the
+          // user and still run the interval to check if the funds were sent.
+          updateState("mintingStep", MintingStep.Deposit)
```
Comment (r-czajkowski) [inline src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx:63]: Let's add `threshold` and `updateState` to the dependency array.

File: src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx [hunk 9]
```diff
@@ -53,19 +73,35 @@ const MintingFlowRouterBase = () => {
...
+
+      if (isDepositRevealed) {
+        updateState("mintingStep", MintingStep.ProvideData)
+        removeDepositDataFromLocalStorage()
+      } else {
+        updateState("mintingStep", MintingStep.InitiateMinting)
+      }
+    }
+    checkIfUtxoIsRevealed()
+  }, [utxo])
```
Comment (r-czajkowski) [inline src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx:92]: Can we do this in the `useEffect` from above? The `utxo` is an object so this effect will be called on every render.

**Review by r-czajkowski — CHANGES_REQUESTED** [review]

File: src/components/Modal/TbtcMintingConfirmationModal/index.tsx (hunk 1 above)
Comment (michalsmiarowski) [reply src/components/Modal/TbtcMintingConfirmationModal/index.tsx]: c848ac024c9ba34761f70c32e32ff9e4c409d14c

File: src/components/Modal/TbtcMintingConfirmationModal/index.tsx (hunk 2 above)
Comment (michalsmiarowski) [reply src/components/Modal/TbtcMintingConfirmationModal/index.tsx]: 0ee520757ac4308dc349d3cfe18b9faae497366f

File: src/components/Modal/TbtcRecoveryFileModal/index.tsx (hunk 3 above)
Comment (michalsmiarowski) [reply src/components/Modal/TbtcRecoveryFileModal/index.tsx:48]: 266552ddac5230ac5bf2a3fca9b125dfc7cc33fa

File: src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts (hunk 5 above)
Comment (michalsmiarowski) [reply src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts]: 2c75eb02480776f97bf9abc28f2cf0c78ce41ae5

File: src/components/Modal/tBTC/GenerateNewDepositAddress.tsx (hunk 4 above)
Comment (michalsmiarowski) [reply src/components/Modal/tBTC/GenerateNewDepositAddress.tsx:31]: See https://github.com/threshold-network/token-dashboard/pull/377#discussion_r1089291288

File: src/pages/tBTC/Bridge/MintingCard/InitiateMinting.tsx (hunk 6 above)
Comment (michalsmiarowski) [reply src/pages/tBTC/Bridge/MintingCard/InitiateMinting.tsx]: 6a9225e58ed310dca65870f173ea700b0b49e43c

File: src/pages/tBTC/Bridge/MintingCard/MakeDeposit.tsx (hunk 7 above)
Comment (michalsmiarowski) [reply src/pages/tBTC/Bridge/MintingCard/MakeDeposit.tsx:42]: 25e3a5a133799f40d7d87a663414e9d6ad20a923

File: src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx (hunk 9 above)
Comment (michalsmiarowski) [reply src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx]: e33c995660a2b9ed0be2efbdf09451abc5032bb0

File: src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx (hunk 8 above)
Comment (michalsmiarowski) [reply src/pages/tBTC/Bridge/MintingCard/MintingFlowRouter.tsx:74]: See https://github.com/threshold-network/token-dashboard/pull/377#discussion_r1089292055

**Review by r-czajkowski — APPROVED** [review]

**Pull request merged**